Update release/PR/issue process documents in devel guide

tpepper opened the pull request:
This PR removes some old documentation, updates and combines some k/sig-release process documentation into the developer guide.

See also https://github.com/kubernetes/sig-release/pull/240

tpepper commented:
/cc @jeefy

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+longer the case.  Features are effectively GitHub issues (or KEPs)
+which lead to subsequent PRs.  The terms feature, issue, and PR may
+be used interchangeably below, but the general labeling process
+should be consistent across artifact types.
```
This is really confusing to me. It seems to say that issues and PRs are the same, and they're really not.

While we may _want to_  label them the same with regard to milestone and priority labeling, this is not currently done. What are the efforts, other than this PR, to encourage folks to do so?
I would be extremely in favor of labeling PRs with these, especially if they are a bug that doesn't have a corresponding k/features issue.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+longer the case.  Features are effectively GitHub issues (or KEPs)
+which lead to subsequent PRs.  The terms feature, issue, and PR may
+be used interchangeably below, but the general labeling process
+should be consistent across artifact types.
```
This is really confusing to me. It seems to say that issues and PRs are the same, and I don't think that's the intent here. I think the explanation below is very clear, and perhaps we can take out the part of this sentence before the comma.

tpepper replied:
Yes.  I'd started to turn the issues.md doc into an issues and PRs doc, but really over the last couple weeks have realized this is a release process doc.  Once I added some info on features, it muddles this sentence.  The aim is the part after the comma...consistent treatment of artifacts associated with the release for simplest user experience.  Workflows are different though for the different artifact types and are given separate sections as of this draft.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+After this point only critical bug fixes happen and all remaining
+issues are resolved before release.
+
```
duplicate information; perhaps prepend with "To summarize code freeze:...."

tpepper replied:
Indeed.  I was trying to better capture this short window dedicated to and needed final stabilization.  I've swizzled words to maybe be more clear.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+issues are resolved before release.
+
+When the code base is sufficiently stable, the master branch re-opens for
+general development and the release is finalize from a branch.
```
typo: "finalized"
clarify this is the end of code freeze and all merges to master will target the _next_ release? maybe I'm overly picky here

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+When the code base is sufficiently stable, the master branch re-opens for
+general development and the release is finalize from a branch.
+
+Following release, the Release Branch Manager cherry picks critical
```
link to Release Branch manager role perhaps?

tpepper replied:
Good thing I have a PR pending to introduce such a thing.  Link added.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+# Targeting Features, Issues and PRs to Release Milestones
+
+This document is focused on kubernetes developers and contributors
```
are we lower-casing kubernetes? It's consistent throughout this doc, but...?

tpepper replied:
https://kubernetes.io/docs/contribute/style/style-guide/#kubernetes-io-word-list says always capitalized.  Fixed throughout doc.  It's not fully consistent elsewhere in the contributor and developer guides.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+## Removal Of Items From The Milestone
+
+Before getting too far into the process for adding, please note:
```
adding features? milestone labels? 

I like the use of "item" to stand in for issue/pr/feature.

tpepper replied:
abstraction FTW

fixed missing words after "for adding, ..."

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+categorize PRs, and to secure support from the relevant SIG to
+guarantee that any breakage caused by the PR will be rapidly resolved.
+
+Where additional action is require, an attempt at human to human
```
typo: "required"

tpepper replied:
fixed

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+channels:
+
+- Comment mentioning the sig team appropriate for the issue type
+- Emailing the SIG googlegroup list
```
- SIGs still have google groups? aren't we moving away from that?
- also perhaps say "mailing list" and link to SIG list?

jeefy replied:
Also nitpick: capitalize sig in "the sig team" to make it uniform with the rest of the doc.

tpepper replied:
Fixed.  While there are still googlegroups, calling them mailing lists is more abstract versus potential future changes in implementatino.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+- Comment mentioning the sig team appropriate for the issue type
+- Emailing the SIG googlegroup list
+  - bootstrapped with the emails addresses from the [community sig list](/sig-list.md)
```
typo: "email"

tpepper replied:
fixed

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+- Emailing the SIG googlegroup list
+  - bootstrapped with the emails addresses from the [community sig list](/sig-list.md)
+  - maybe configured to email alternate googlegroup
+  - maybe configured to directly email SIG leads or other SIG members
```
I don't understand these three bullet points. Are they explaining potential different ways in which a google group is set up?

tpepper replied:
Clarified.  This was definitely oddly worded in the old doc.  My read is "configured to" means "addressed to".  So one might either address "To:" the group or direct some specific user@domain.com when writing an escalation email.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+  - maybe configured to directly email SIG leads or other SIG members
+- Messaging the SIG slack channel, mentioning the SIG leads
+  - bootstrapped with the slackchannel and SIG leads from the [community sig list](/sig-list.md)
+  - maybe configured to message alternate slack channel and users
```
ditto; I don't entirely understand this information. If a chain reaction of notifications is caused, I'm not even sure it belongs in this document - the most important point is that messages are sent and received. SIGs configuring their information channels is really up to them.

tpepper replied:
Yeah this was all straight from the prior document, which was a mix of:
* letting the release team know what it should do relative to the dev community
* letting the dev membership understand what to expect from the release team

tpepper replied:
I think it's reasonable, especially in the dev guide now to give a sense of, "Hey if you see the release team folks on about something..they're doing their escalation job."  Maybe something direct and informal like that conveys the sentiment better?

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+In the past there was a requirement for a milestone targeted pull
+request to have an associated GitHub issue opened, but this is no
+longer the case.  Features are effectively GitHub issues (or KEPs)
```
perhaps link to KEP process, at least once?

tpepper replied:
added

spiffxp replied:
the parenthetical makes it look like like issues are KEPs

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+Feature planning and definition takes many forms today, but a typical
+example might be a large piece of work described in a KEP, with associated
+task issues in GitHub.  When the plan has reach an implementable state and
```
typo: "reached"

tpepper replied:
fixed

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+
+For the first ~4 weeks into the release cycle, the release team's
+Features Lead will interact with SIGs and feature owners via GitHub,
+Slack, and SIG meetings to captur all required planning artifacts.
```
typo: "capture"

tpepper replied:
fixed

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+Issues are marked as targeting a milestone via the Prow
+"/milestone" command.
+
+The release team's Bug Triage Lead and overall community watch
```
link to Bug Triage Lead description?

tpepper replied:
added

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+The SIG owner label defines the SIG to which we escalate if a
+milestone issue is languishing or needs additional attention.  If
+there are no updates after escalation, the issue may be automatically
+removed from the milestone.
```
perhaps give example of SIG owner label to dissociate from OWNERS files
(in my head I only think of it as a "SIG label")

tpepper replied:
Hah.  That was there in the original issues.md and I deleted it thinking it looked redundant.  Re-added for clarity.

guineveresaenger commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+- `kind/bug`: Fixes a newly discovered bug.
+  - were not known issues at the start of the development period.
+- `kind/feature`: New functionality.
+- `kind/cleanup`: Adding tests, refactoring, fixing old bugs.
```
Add `kind/failing-test` to help CI lead

tpepper replied:
Added, along with flake.

jeefy commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+project scale and globally distributed developer base, it is critical
+to project velocity that continuous integration testing insures the
+project is continuous stable so that individual commits can be
+flagged as having broken something.
```
1) I'd capitalize each of the phases.
2) Fix formatting by adding another new line after stabilization

tpepper replied:
fixed

jeefy commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+* implementation
+* stabilization
+But in reality this is an open source and agile project, with feature
+planning and implementation happening at all times.  And given the
```
Might be nit-picky but I'd just remove "And" and capitalize "given".

MHBauer commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+But in reality this is an open source and agile project, with feature
+planning and implementation happening at all times.  And given the
+project scale and globally distributed developer base, it is critical
+to project velocity that continuous integration testing insures the
```
insures -> ensures

MHBauer commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+planning and implementation happening at all times.  And given the
+project scale and globally distributed developer base, it is critical
+to project velocity that continuous integration testing insures the
+project is continuous stable so that individual commits can be
```
continuous -> continuously

MHBauer commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,231 @@
+Implementation and bugfixing is ongoing across the cycle, but
+culminates in a code slush and code freeze period:
+* The **code slush** starts in week ~9 of the release cycle.  The master
+  branch only accepts PRs for release milestone.  No additional feature
```
for -> for a

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+    -   [Other Required Labels](#other-required-labels)
+        -   [SIG Owner Label](#sig-owner-label)
+        -   [Priority Label](#priority-label)
+        -   [Issue Kind Label](#issue-kind-label)
```
The shape of the nesting looks weird. I also recall discussions about wanting better label hygiene.  Maybe pull "Other Required Labels" into a top level point?

tpepper replied:
fixed

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+* the feature, issue, or pull request owner
+* SIG leadership
+* the release team
+Information on workflows and interactions are described below.
```
nit: linebreak between the list and this sentence

tpepper replied:
fixed

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+* the release team
+Information on workflows and interactions are described below.
+
+As the owner of a feature, issue, or pull request (PR), it is your
```
What's my tl;dr here?  I appreciate there may be a lot of context below, but you're going to lose me in all of those paragraphs

spiffxp replied:
I think the tl;dr I'm looking for is something like "what do I have to do to get my code into the upcoming release"
- Initially you must frob the whatsit
- During code freeze you must flip the doodad
- After code freeze you must perform an interpretive dance for the following audience

There also seems to be an undefined implication that being "in the milestone" is the same that as being "in the release".  That's not really true today, there are tons of PR's being merged right that have no milestone attached.

guineveresaenger replied:
this state of things does beg the question on what is _required_ for a milestone/why we have a milestone process at all.

tpepper replied:
added TL;DR table

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+  contact info for the team associated with any given release can be
+  found [here](https://git.k8s.io/sig-release/releases/).
+- *Y days*: Refers to business days (using the location local to the release-manager M-F).
+
```
What is a "feature"

spiffxp replied:
What is a "release milestone"

tpepper replied:
added reference info

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+# Targeting Features, Issues and PRs to Release Milestones
+
+This document is focused on Kubernetes developers and contributors
```
More specifically I think we're just talking about code that is intended to land in kubernetes/kubernetes?  How does this process handle repos like kubernetes/kubeadm?

tpepper replied:
This is an interesting question and currently one without an answer in my opinion.  I don't see it being answered or dictated by this document.  If the document in the meantime gives a description of k/k practice which is embraced by other repos, it would be a positive change toward common experience across the project.

If folks REALLY want the document constrained I could change "Kubernetes developers and contributors" to "kubernetes/kubernetes repo developers and contributors".  But I'll be sad to do it ;)

spiffxp replied:
Fair, let's leave it as is and see what happens

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+
+The members of the GitHub [“kubernetes-milestone-maintainers”
+team](https://github.com/orgs/kubernetes/teams/kubernetes-milestone-maintainers/members)
+are entrusted with the responsibility of specifying milestone tags
```
are "milestone tags" the same thing as "release milestones"?

tpepper replied:
Changed the wording to: "...specifying the Milestone on
GitHub artifacts."

I've partly tried to avoid being too GitHub process specific, but I guess this is inevitable.  We have overlapping and overloaded nouns.

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+[KEP](https://git.k8s.io/community/keps), with associated
+task issues in GitHub.  When the plan has reached an implementable state and
+work is underway, the feature or parts thereof are targeted for an upcoming
+milestone by creating GitHub issues and marking them with the Prow /milestone
```
Style nit: you use "/milestone" command elsewhere

nikhita replied:
Here, GitHub issues = issues on k/k or issues on k/features?

tpepper replied:
¯\_(ツ)_/¯  Yes?

SIG-PM is working towards what should be a KEP and clarify this.

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+
+An open issue is no longer required for a PR, but open issues and
+associated PRs should have synchronized labels.  For example a high
+priority bug issue might not have its associated PR merged if the PR is
```
is this "in the event that issue labels != pr labels, release team behavior is undefined" or does one take precedence over the other?

tpepper replied:
Yes.  This is an undefined situation in current practice and process.

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+
+Marking issues with the milestone provides the community better
+visibility regarding when an issue was observed and by when the community
+feels it must be resolved.
```
when are issues required to have a milestone vs. it being nice to have add'l visibility?

tpepper replied:
Added at end of paragraph "During code freeze a Milestone is PR merge requirement."

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+"/milestone" command.
+
+This is a blocking requirement during code slush and code freeze as
+described above.
```
(mostly for my understanding, not suggesting wording here) so the corollary is that milestones on PR's are purely nice to have until code slush... we're not going to dissuade their use, but we're not really going to pay attention to them either 

(this is where I believe @jdumars was hoping we could one day move to milestones always being required or auto-added by the bot, to better help us measure/compare the volume of work during burndown (slush+freeze+post-freeze) relative to pre-burndown)

tpepper replied:
Yes, your assessment match my take on things currently.

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+This is a blocking requirement during code slush and code freeze as
+described above.
+
+### Other Required Labels
```
Are these required on issues or PR's? Or both?

tpepper replied:
Only hard requirement I'm aware of today is that enforced by PR merge automation.  There are non-mechanically enforced "desires" on consistency across feature issues, issues and PRs and across repos for that matter too.  Again awaiting SIG-PM proposal.

tpepper commented:
Great feedback and questions @spiffxp.  Will fix up the formatty ones later today.

There's definitely a need for a TL;DR also.  I'm not sure what form this should take.  For example:
* @castrojo 's suggested a dashboard somewhere indicating summarizing what phase we're in, merge requirements, and maybe some metric of health.  Could be linked from this doc.
* the old https://github.com/kubernetes/sig-release/blob/master/ephemera/issues.md in a way gave the summary of requirements through negation in terms of what the bot did to remove your stuff, but it was sooooo long as to not really be readable
* maybe a small table at the top of the document is viable?

I've biased in this initial writing toward describing somewhat abstractly the process to give a sense of the "why" story.  That's partly acknowledging that the specific "what" are in flux right now for 1.12 and there's talk from SIG-PM about more overarching changes too.  So this document if it gets very specific is potentially out of date in a week.  This makes me like the idea of dashboard, fed by the same yaml that feeds the automation that does enforcing.

In the meantime a table maybe like:

|          | *Weeks 1-8* | *Week 9* | *Weeks 10-12* | *Weeks 12+* |             
| *Required Labels* | <ul> <li>/milestone (if feature)</li> <li>/sig {name}</li> <li>/kind {type}</li> <li>/lgtm</li> <li>/approved</li> </ul> | <ul> <li>/milestone</li> <li>/sig {name}</li> <li>/kind {type}</li> <li>/priority {level}</li> <li>/lgtm</li> <li>/approved</li> </ul> | <ul> <li>/milestone</li> <li>/sig {name}</li> <li>/kind {bug, failing-test}</li> <li>/priority critical-urgent</li> <li>/lgtm</li> <li>/approved</li> </ul> | <ul> <li>/milestone</li> <li>/sig {name}</li> <li>/kind {bug, failing-test}</li> <li>/priority critical-urgent</li> <li>/lgtm</li> <li>/approved</li> <li>cherry-pick labels set by tooling and release branch patch manager</li> </ul> |

nikhita commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+Implementation and bugfixing is ongoing across the cycle, but
+culminates in a code slush and code freeze period:
+* The **code slush** starts in week ~9 of the release cycle.  The master
+  branch only accepts PRs for a release milestone.  No additional feature
```
nit: **the** release milestone?

tpepper replied:
Good clarification!  Changed "a" to "the upcoming".

nikhita commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+culminates in a code slush and code freeze period:
+* The **code slush** starts in week ~9 of the release cycle.  The master
+  branch only accepts PRs for a release milestone.  No additional feature
+  development is merged after this point.
```
after this point = during code slush i.e. in week 9?

I've seen PRs with features (with the correct labels and milestone set) being merged just before the code freeze.

tpepper replied:
I don't see this able to be more than a cultural request to project leadership to not merge late features, without moving to a high friction, high latency project management process and enforcing mechanisms.  So I'm trying to capture policy intent as it stands today, even if that isn't always matched by human actions or enforced by mechanism.

nikhita replied:
> So I'm trying to capture policy intent as it stands today, even if that isn't always matched by human actions or enforced by mechanism.

:+1:

nikhita commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,246 @@
+When the code base is sufficiently stable, the master branch re-opens
+for general development and work begins there for the next release
+milestone. Any remaining modifications for the current release are cherry
+picked from master back to the release branch.  The release is built from
```
Can you also add when the release branch for the current release is created? I'm not sure if that's really required for a contributor to know but it helps with the flow/reading. :)

tpepper replied:
Added a definition.

nikhita commented:
/sig release

guineveresaenger commented:
I love the table as well!

tpepper commented:
Worth noting the old doc said https://github.com/kubernetes/sig-release/blob/master/ephemera/issues.md#issues-tracked-in-other-repos on the topic of other repos.  I've ended up dropping that section in its entirety, but don't have anything concrete to assert in its place in advance of what's coming through SIG-PM for change proposals.

jeefy commented:
That table is gold!

tpepper commented:
/hold cancel

I believe I've addressed all the comments here as best is possible today.  There are definitely some big picture process questions that will yet evolve in the community over time, but I'm not expecting those to change in 1.12's timeline.

jdumars commented:
/approve

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,315 @@
+target milestone is a prior release (see [cherry pick
+process](cherry-picks.md) for more information).
+
+## TL;DR
```
This is really close, but I'm still a little confused:
- what is this table describing? ie: "tl;dr If you want your PR to get merged, it needs the following required labels and milestones, represented here by the `/commands` it would take to add them"
- is this for PR's only, or also for issues?
- ... because issues don't need /lgtm and /approved

tpepper replied:
Yes.  Added hopefully clarifying text.

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,315 @@
+<td>Required Labels</td>
+<td>
+<ul>
+<li>/milestone (if feature)</li>
```
`/milestone {v1.y}` (ditto for the rest)

tpepper replied:
added

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,315 @@
+<li>/priority critical-urgent</li>
+<li>/lgtm</li>
+<li>/approved</li>
+<li>cherry-pick labels set by tooling and release branch patch manager</li>
```
~by tooling and~

tpepper replied:
removed

spiffxp commented on contributors/devel/release.md:
```diff
@@ -0,0 +1,315 @@
+
+The members of the GitHub [“kubernetes-milestone-maintainers”
+team](https://github.com/orgs/kubernetes/teams/kubernetes-milestone-maintainers/members)
+are entrusted with the responsibility of specifying the Milestone on
```
nit: this is the only place you capitalize Milestone, maybe s/Milestone/release-milestone/

tpepper replied:
Yeah the wording here is confusing.  It feels like a label, but it's a first class GitHub entity and we use "release" and "milestone" interchangeably in so many places.  I found a couple other places with 'M'.  GitHub doesn't treat it as a proper noun.  I've swizzled them now into "release milestone".

spiffxp commented:
/lgtm
Thanks for the thoughtful iteration
